sdk/python/lib/pulumi/runtime/rpc.py

```diff
@@ -171,7 +181,14 @@ def deserialize_properties(props_struct: struct_pb2.Struct) -> Any:
         raise AssertionError("Unrecognized signature when unmarshaling resource property")
 
     # Struct is duck-typed like a dictionary, so we can iterate over it in the normal ways.
-    return {k: deserialize_property(v) for (k, v) in list(props_struct.items())}
+    output = {}
+    for k, v in list(props_struct.items()):
+        value = deserialize_property(v)
+        # We treat values that deserialize to "None" as if they don't exist.
+        if value is not None:
```

Comment:
This code (and the corresponding serialization code) is also subtle.

`deserialize_property`, when called on the `UNKNOWN` sentinel, produces None. If value is `None`, that implies that it was unknown to the engine. By not adding it to the output dictionary, when we resolve all of the outstanding futures on a resource when `RegisterResource` returns, we'll report that an output is known if its value is not None. So, from this direction, everything works fine - user code never observes UNKNOWN because applies on unknown outputs don't run during previews.

From the serialization direction, all resources initialize their output properties to the value `None`, so if they are not provided they are not known to the SDK when calling `RegisterResource` and it does not make sense to include them in the `RegisterResource` RPC call.

Reply:
It's consistent with Node's behavior. Python's previous behavior was to directly expose unknowns in the programming model using a special `Unknown` type that overloaded a bunch of operators to masquerade as a string.

Reply:
> It's consistent with Node's behavior

Modulo the fact that Python is conflating `null` and `unknown`, right? In practice I don't think that this is a big deal: I'd be very surprised if we ever encounter `null` values in the field.

Reply:
(we should investigate removing `null` from the Pulumi type system tbqh)

Reply:
> Modulo the fact that Python is conflating null and unknown, right?

Yeah, true. But yeah, I agree that this is probably not a big deal.